smtp/client.ts

```diff
@@ -1,87 +1,53 @@
-const { SMTP, state } = require('./smtp');
-const { Message, create } = require('./message');
-const addressparser = require('addressparser');
+import addressparser from 'addressparser';
+import { Message } from './message';
+import type { MessageAttachment, MessageHeaders } from './message';
+import { SMTPConnection, SMTPState } from './smtp';
+import type { SMTPConnectionOptions } from './smtp';
+
+export interface MessageStack {
+	callback: (error: Error | null, message: Message) => void;
+	message: Message;
+	attachment: MessageAttachment;
+	text: string;
+	returnPath: string;
+	from: string;
+	to: ReturnType<typeof addressparser>;
+	cc: string[];
+	bcc: string[];
+}
+
+export class Client {
+	public readonly smtp: SMTPConnection;
+	public readonly queue: MessageStack[] = [];
+
+	protected sending = false;
+	protected ready = false;
+	protected timer: NodeJS.Timer | null = null;
 
-class Client {
 	/**
-	 * @typedef {Object} MessageStack
-	 * @property {function(Error, Message): void} [callback]
-	 * @property {Message} [message]
-	 * @property {string} [returnPath]
-	 * @property {string} [from]
-	 * @property {string} [subject]
-	 * @property {string|Array} [to]
-	 * @property {Array} [cc]
-	 * @property {Array} [bcc]
-	 * @property {string} [text]
-	 * @property {*} [attachment]
-	 *
-	 * @typedef {Object} SMTPSocketOptions
-	 * @property {string} key
-	 * @property {string} ca
-	 * @property {string} cert
-	 *
-	 * @typedef {Object} SMTPOptions
-	 * @property {number} [timeout]
-	 * @property {string} [user]
-	 * @property {string} [password]
-	 * @property {string} [domain]
-	 * @property {string} [host]
-	 * @property {number} [port]
-	 * @property {boolean|SMTPSocketOptions} [ssl]
-	 * @property {boolean|SMTPSocketOptions} [tls]
-	 * @property {string[]} [authentication]
-	 * @property {function(...any): void} [logger]
-	 *
-	 * @constructor
-	 * @param {SMTPOptions} server smtp options
+	 * @param {SMTPConnectionOptions} server smtp options
```

Comment:
the inline docs get copied into the bundle, which helps give (for lack of a better term) commodity-level code completion support in scenarios where you wouldn't have access to the original ts files alongside the bundle

i say "commodity-level" because they're not bulletproof; all our custom interfaces in particular get erased at compile time, so those types will still be holes. but it's "better than nothing" at least.